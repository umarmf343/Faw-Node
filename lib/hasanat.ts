const ARABIC_LETTER_REGEX = /[\u0621-\u063A\u0641-\u064A]/g

export function countArabicLetters(text: string): number {
  if (!text) {
    return 0
  }
  return text.match(ARABIC_LETTER_REGEX)?.length ?? 0
}

export function calculateHasanatForText(text: string, rewardPerLetter = 10): number {
  const letters = countArabicLetters(text)
  return letters * rewardPerLetter
}
