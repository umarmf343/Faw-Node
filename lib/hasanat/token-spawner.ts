"use client"

export type HasanatTokenMode = "float" | "drift" | "reduced"

export type TokenRect = Pick<DOMRectReadOnly, "x" | "y" | "width" | "height" | "top" | "right" | "bottom" | "left">

export interface HasanatTokenModel {
  id: string
  amount: number
  originRect: TokenRect
  targetRect?: TokenRect
  createdAt: number
  mode: HasanatTokenMode
}

type TokenListener = (tokens: HasanatTokenModel[]) => void

const AGGREGATION_WINDOW_MS = 320
const MAX_TOKENS = 5
const DRIFT_DURATION_MS = 820
const FLOAT_GUARD_MS = 1500

const isServer = () => typeof window === "undefined"

function createRect(rect: DOMRectReadOnly): TokenRect {
  return {
    x: rect.x,
    y: rect.y,
    width: rect.width,
    height: rect.height,
    top: rect.top,
    right: rect.right,
    bottom: rect.bottom,
    left: rect.left,
  }
}

function createId() {
  if (typeof crypto !== "undefined" && typeof crypto.randomUUID === "function") {
    return crypto.randomUUID()
  }
  return `token_${Date.now()}_${Math.random().toString(16).slice(2)}`
}

function isRectVisible(rect: TokenRect) {
  if (isServer()) {
    return false
  }
  const viewportHeight = window.innerHeight || 0
  const viewportWidth = window.innerWidth || 0
  return rect.bottom >= 0 && rect.top <= viewportHeight && rect.right >= 0 && rect.left <= viewportWidth
}

function isRectOnScreen(rect: TokenRect) {
  return rect.width > 0 && rect.height > 0 && isRectVisible(rect)
}

class TokenSpawner {
  private tokens: HasanatTokenModel[] = []
  private listeners = new Set<TokenListener>()
  private cleanupTimers = new Map<string, ReturnType<typeof setTimeout>>()
  private aggregateTokenId: string | null = null
  private aggregateTimer: ReturnType<typeof setTimeout> | null = null
  private prefersReducedMotion = false
  private reducedMotionOverride: boolean | null = null
  private motionMedia?: MediaQueryList
  private motionListener?: (event: MediaQueryListEvent) => void
  private lastOriginRect?: TokenRect

  constructor() {
    if (isServer()) {
      return
    }

    if (typeof window.matchMedia === "function") {
      const media = window.matchMedia("(prefers-reduced-motion: reduce)")
      this.prefersReducedMotion = media.matches
      const handle = (event: MediaQueryListEvent) => {
        this.setReducedMotionPreference(event.matches)
      }

      if (typeof media.addEventListener === "function") {
        media.addEventListener("change", handle)
      } else if (typeof media.addListener === "function") {
        media.addListener(handle)
      }

      this.motionMedia = media
      this.motionListener = handle
    }
  }

  subscribe(listener: TokenListener) {
    this.listeners.add(listener)
    listener(this.tokens)
    return () => {
      this.listeners.delete(listener)
    }
  }

  spawn(amount = 1) {
    if (isServer() || amount <= 0) {
      return
    }

    const origin = this.measureOriginRect()
    if (!origin) {
      return
    }

    this.lastOriginRect = origin

    const { mode, targetRect } = this.resolveMode(origin)

    const now = performance.now()

    if (this.aggregateTokenId) {
      const aggregateToken = this.tokens.find((token) => token.id === this.aggregateTokenId)
      if (aggregateToken && now - aggregateToken.createdAt <= AGGREGATION_WINDOW_MS) {
        const updated: HasanatTokenModel = {
          ...aggregateToken,
          amount: aggregateToken.amount + amount,
          createdAt: now,
          mode,
          originRect: origin,
          targetRect: mode === "drift" ? targetRect : undefined,
        }
        this.tokens = this.tokens.map((token) => (token.id === updated.id ? updated : token))
        this.scheduleAggregateReset()
        this.scheduleCleanup(updated.id)
        this.notify()
        return
      }
    }

    const token: HasanatTokenModel = {
      id: createId(),
      amount,
      originRect: origin,
      targetRect: mode === "drift" ? targetRect : undefined,
      createdAt: now,
      mode,
    }

    this.tokens = [...this.tokens.slice(-MAX_TOKENS + 1), token]
    this.aggregateTokenId = token.id
    this.scheduleAggregateReset()
    this.scheduleCleanup(token.id)
    this.notify()
  }

  complete(tokenId: string) {
    if (!tokenId) {
      return
    }

    this.tokens = this.tokens.filter((token) => token.id !== tokenId)
    this.clearCleanup(tokenId)
    if (this.aggregateTokenId === tokenId) {
      this.aggregateTokenId = null
      this.clearAggregateTimer()
    }
    this.notify()
  }

  triggerCounterPulse(reduced = false) {
    if (isServer()) {
      return
    }

    const target = this.findCounterElement()
    if (!target) {
      return
    }

    const pulseClass = reduced ? "animate-hasanat-counter-reduced" : "animate-hasanat-counter"
    target.classList.remove(pulseClass)
    // Force reflow to restart animation
    // eslint-disable-next-line @typescript-eslint/no-unused-expressions
    target.clientHeight
    target.classList.add(pulseClass)

    window.setTimeout(() => {
      target.classList.remove(pulseClass)
    }, reduced ? 220 : 260)
  }

  reset() {
    this.tokens = []
    this.notify()
    this.cleanupTimers.forEach((timer) => window.clearTimeout(timer))
    this.cleanupTimers.clear()
    this.aggregateTokenId = null
    this.clearAggregateTimer()
  }

  setReducedMotionPreference(value: boolean) {
    this.prefersReducedMotion = value
  }

  setReducedMotionOverride(value: boolean | null) {
    this.reducedMotionOverride = value
  }

  private get isReducedMotion() {
    return this.reducedMotionOverride ?? this.prefersReducedMotion
  }

  private measureOriginRect(): TokenRect | undefined {
    if (isServer()) {
      return undefined
    }

    const candidates = Array.from(
      document.querySelectorAll<HTMLElement>("[data-hasanat-source=\"next-button\"]"),
    )

    for (const element of candidates) {
      if (element instanceof HTMLButtonElement && element.disabled) {
        continue
      }
      const rect = createRect(element.getBoundingClientRect())
      if (isRectOnScreen(rect)) {
        return rect
      }
    }

    return this.lastOriginRect
  }

  private resolveMode(originRect: TokenRect) {
    if (this.isReducedMotion) {
      return { mode: "reduced" as const }
    }

    const target = this.measureCounterRect()
    if (target) {
      return { mode: "drift" as const, targetRect: target }
    }

    return { mode: "float" as const }
  }

  private measureCounterRect(): TokenRect | undefined {
    if (isServer()) {
      return undefined
    }

    const element = this.findCounterElement()
    if (!element) {
      return undefined
    }

    const rect = createRect(element.getBoundingClientRect())
    if (!isRectOnScreen(rect)) {
      return undefined
    }
    return rect
  }

  private findCounterElement() {
    if (isServer()) {
      return null
    }
    return document.querySelector<HTMLElement>("[data-hasanat-target=\"hasanat-counter\"]")
  }

  private scheduleCleanup(tokenId: string) {
    this.clearCleanup(tokenId)
    const timer = window.setTimeout(() => {
      this.complete(tokenId)
    }, Math.max(FLOAT_GUARD_MS, DRIFT_DURATION_MS + 200))
    this.cleanupTimers.set(tokenId, timer)
  }

  private clearCleanup(tokenId: string) {
    const timer = this.cleanupTimers.get(tokenId)
    if (timer) {
      window.clearTimeout(timer)
      this.cleanupTimers.delete(tokenId)
    }
  }

  private scheduleAggregateReset() {
    this.clearAggregateTimer()
    this.aggregateTimer = window.setTimeout(() => {
      this.aggregateTokenId = null
    }, AGGREGATION_WINDOW_MS + 40)
  }

  private clearAggregateTimer() {
    if (this.aggregateTimer) {
      window.clearTimeout(this.aggregateTimer)
      this.aggregateTimer = null
    }
  }

  private notify() {
    const snapshot = this.tokens.map((token) => ({ ...token }))
    this.listeners.forEach((listener) => listener(snapshot))
  }
}

export const tokenSpawner = new TokenSpawner()
