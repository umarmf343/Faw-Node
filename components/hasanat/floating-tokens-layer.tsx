"use client"

import * as React from "react"
import { createPortal } from "react-dom"

import { useUser } from "@/hooks/use-user"
import type { HasanatTokenModel, TokenRect } from "@/lib/hasanat/token-spawner"
import { tokenSpawner } from "@/lib/hasanat/token-spawner"

const DRIFT_DURATION = 820
const FLOAT_DURATION = 900
const REDUCED_DURATION = 260

function getCenter(rect: TokenRect) {
  return {
    x: rect.left + rect.width / 2,
    y: rect.top + rect.height / 2,
  }
}

function easeOutCubic(t: number) {
  return 1 - Math.pow(1 - t, 3)
}

function getQuadraticPoint(start: { x: number; y: number }, control: { x: number; y: number }, end: {
  x: number
  y: number
}, t: number) {
  const oneMinusT = 1 - t
  const x = oneMinusT * oneMinusT * start.x + 2 * oneMinusT * t * control.x + t * t * end.x
  const y = oneMinusT * oneMinusT * start.y + 2 * oneMinusT * t * control.y + t * t * end.y
  return { x, y }
}

function getControlPoint(start: { x: number; y: number }, end: { x: number; y: number }) {
  const midX = (start.x + end.x) / 2
  const horizontalShift = (end.x - start.x) * 0.18
  const verticalLift = Math.max(64, Math.abs(start.y - end.y) * 0.35)
  const controlY = Math.min(start.y, end.y) - verticalLift
  return {
    x: midX - horizontalShift,
    y: controlY,
  }
}

function FloatingToken({ token, prefersArabic }: { token: HasanatTokenModel; prefersArabic: boolean }) {
  const containerRef = React.useRef<HTMLDivElement>(null)
  const contentRef = React.useRef<HTMLDivElement>(null)

  React.useEffect(() => {
    const container = containerRef.current
    const content = contentRef.current
    if (!container || !content) {
      return
    }

    const startPoint = getCenter(token.originRect)
    container.style.transform = `translate3d(${startPoint.x}px, ${startPoint.y}px, 0)`
    content.style.opacity = "0"
    content.style.transform = "translate(-50%, -50%) scale(0.85)"
    content.style.contain = "paint"

    if (token.mode === "float") {
      const lift = Math.min(96, Math.max(56, token.originRect.height * 1.45))
      content.style.setProperty("--hasanat-float-distance", `${lift}px`)
      content.style.setProperty("--hasanat-float-duration", `${FLOAT_DURATION}ms`)
      content.classList.remove("animate-hasanat-token-float")
      void content.offsetWidth
      content.classList.add("animate-hasanat-token-float")
      const timeoutId = window.setTimeout(() => {
        tokenSpawner.complete(token.id)
      }, FLOAT_DURATION + 160)
      return () => {
        window.clearTimeout(timeoutId)
        content.classList.remove("animate-hasanat-token-float")
      }
    }

    if (token.mode === "reduced") {
      const duration = REDUCED_DURATION
      content.style.transition = `opacity ${duration}ms ease-out`
      requestAnimationFrame(() => {
        content.style.opacity = "1"
        window.setTimeout(() => {
          content.style.opacity = "0"
        }, duration - 90)
      })
      const timeoutId = window.setTimeout(() => {
        tokenSpawner.triggerCounterPulse(true)
        tokenSpawner.complete(token.id)
      }, duration + 120)
      return () => {
        window.clearTimeout(timeoutId)
        content.style.transition = ""
      }
    }

    if (!token.targetRect) {
      const timeoutId = window.setTimeout(() => tokenSpawner.complete(token.id), FLOAT_DURATION + 180)
      return () => {
        window.clearTimeout(timeoutId)
      }
    }

    const endPoint = getCenter(token.targetRect)
    const controlPoint = getControlPoint(startPoint, endPoint)
    let frame = 0
    let startTime: number | null = null

    const step = (timestamp: number) => {
      if (startTime === null) {
        startTime = timestamp
      }
      const elapsed = timestamp - startTime
      const progress = Math.min(elapsed / DRIFT_DURATION, 1)
      const eased = easeOutCubic(progress)
      const position = getQuadraticPoint(startPoint, controlPoint, endPoint, eased)
      const scale = progress < 0.2 ? 0.85 + (progress / 0.2) * 0.15 : 1
      const opacity =
        progress < 0.15 ? progress / 0.15 : progress > 0.8 ? 1 - (progress - 0.8) / 0.2 : 1

      container.style.transform = `translate3d(${position.x}px, ${position.y}px, 0)`
      content.style.transform = `translate(-50%, -50%) scale(${scale})`
      content.style.opacity = `${opacity}`

      if (progress < 1) {
        frame = requestAnimationFrame(step)
      } else {
        tokenSpawner.triggerCounterPulse(false)
        tokenSpawner.complete(token.id)
      }
    }

    frame = requestAnimationFrame(step)
    return () => {
      cancelAnimationFrame(frame)
    }
  }, [token])

  return (
    <div ref={containerRef} className="pointer-events-none absolute left-0 top-0 will-change-transform">
      <div
        ref={contentRef}
        className="pointer-events-none select-none rounded-full bg-emerald-500/90 px-3 py-1.5 text-xs font-semibold text-white shadow-[0_16px_40px_rgba(16,185,129,0.35)] ring-1 ring-emerald-300/35 backdrop-blur-sm dark:bg-emerald-400/90 dark:text-emerald-950"
        aria-hidden="true"
      >
        <div className="flex flex-col items-center gap-0.5 leading-none">
          <span className="text-base font-semibold tracking-tight">
            +{token.amount.toLocaleString()}
          </span>
          <span className="text-[0.6rem] font-medium uppercase tracking-[0.35em] text-emerald-50/80 dark:text-emerald-900/70">
            Hasanāt
          </span>
          {prefersArabic ? (
            <span className="text-[0.65rem] font-medium text-emerald-100/80 dark:text-emerald-900/70">
              حسنات
            </span>
          ) : null}
        </div>
      </div>
    </div>
  )
}

export function FloatingTokensLayer() {
  const { stats, profile } = useUser()
  const [tokens, setTokens] = React.useState<HasanatTokenModel[]>([])
  const [mounted, setMounted] = React.useState(false)
  const [announcement, setAnnouncement] = React.useState("")

  const prefersArabic = React.useMemo(() => {
    const locale = profile?.locale ?? ""
    return locale.toLowerCase().startsWith("ar")
  }, [profile?.locale])

  React.useEffect(() => {
    setMounted(true)
  }, [])

  React.useEffect(() => {
    const unsubscribe = tokenSpawner.subscribe(setTokens)
    return () => {
      unsubscribe()
    }
  }, [])

  React.useEffect(() => {
    if (tokens.length === 0) {
      return
    }
    const latest = tokens[tokens.length - 1]
    setAnnouncement(`Hasanāt +${latest.amount.toLocaleString()}. Total: ${stats.hasanat.toLocaleString()}.`)
  }, [stats.hasanat, tokens])

  if (!mounted || typeof document === "undefined") {
    return null
  }

  return createPortal(
    <div
      data-testid="hasanat-token-layer"
      className="pointer-events-none fixed inset-0 z-[70] overflow-visible"
    >
      <div aria-live="polite" role="status" className="sr-only">
        {announcement}
      </div>
      {tokens.map((token) => (
        <FloatingToken key={`${token.id}-${token.createdAt}`} token={token} prefersArabic={prefersArabic} />
      ))}
    </div>,
    document.body,
  )
}
