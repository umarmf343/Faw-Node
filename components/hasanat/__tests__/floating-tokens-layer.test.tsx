import React from "react"
import { render, cleanup, waitFor } from "@testing-library/react"
import { describe, expect, afterEach, beforeEach, it, vi } from "vitest"

import type { HasanatTokenModel, TokenRect } from "@/lib/hasanat/token-spawner"

import { FloatingTokensLayer } from "@/components/hasanat/floating-tokens-layer"

let mockTokens: HasanatTokenModel[] = []

vi.mock("@/lib/hasanat/token-spawner", () => {
  const listeners = new Set<(tokens: HasanatTokenModel[]) => void>()
  return {
    tokenSpawner: {
      subscribe: (listener: (tokens: HasanatTokenModel[]) => void) => {
        listeners.add(listener)
        listener(mockTokens)
        return () => {
          listeners.delete(listener)
        }
      },
      complete: vi.fn(),
      triggerCounterPulse: vi.fn(),
      setReducedMotionOverride: vi.fn(),
      reset: vi.fn(),
      spawn: vi.fn(),
    },
  }
})
const mockUser = {
  stats: { hasanat: 128 },
  profile: { locale: "en-US" },
}

vi.mock("@/hooks/use-user", () => ({
  useUser: () => mockUser,
}))

function createTokenRect(x: number, y: number, width: number, height: number): TokenRect {
  return {
    x,
    y,
    width,
    height,
    top: y,
    left: x,
    right: x + width,
    bottom: y + height,
  }
}

describe("FloatingTokensLayer", () => {
  beforeEach(() => {
    mockTokens = []
    mockUser.profile.locale = "en-US"
  })

  afterEach(() => {
    cleanup()
  })

  it("renders float and drift tokens with announcement", async () => {
    mockTokens = [
      {
        id: "float-1",
        amount: 1,
        mode: "float",
        originRect: createTokenRect(180, 540, 80, 40),
        createdAt: 1,
      },
      {
        id: "drift-1",
        amount: 3,
        mode: "drift",
        originRect: createTokenRect(200, 560, 80, 40),
        targetRect: createTokenRect(360, 90, 120, 40),
        createdAt: 2,
      },
    ]

    render(<FloatingTokensLayer />)

    await waitFor(() => {
      expect(document.querySelector("[data-testid='hasanat-token-layer']")).not.toBeNull()
    })

    const layer = document.querySelector("[data-testid='hasanat-token-layer']") as HTMLElement
    expect(layer).toMatchSnapshot()
    const status = layer.querySelector("[role='status']")
    expect(status?.textContent).toContain("Hasanāt +3. Total: 128.")
  })

  it("shows Arabic accent labels when locale is Arabic", async () => {
    mockUser.profile.locale = "ar-SA"
    mockTokens = [
      {
        id: "reduced-1",
        amount: 2,
        mode: "reduced",
        originRect: createTokenRect(200, 560, 80, 40),
        createdAt: 3,
      },
    ]

    render(<FloatingTokensLayer />)

    await waitFor(() => {
      expect(document.querySelector("[data-testid='hasanat-token-layer']")).not.toBeNull()
    })

    const layer = document.querySelector("[data-testid='hasanat-token-layer']") as HTMLElement
    expect(layer.textContent).toContain("حسنات")
    expect(layer).toMatchSnapshot()
  })
})
