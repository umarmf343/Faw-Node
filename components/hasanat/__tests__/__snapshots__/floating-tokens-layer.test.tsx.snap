// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FloatingTokensLayer > renders float and drift tokens with announcement 1`] = `
<div
  class="pointer-events-none fixed inset-0 z-[70] overflow-visible"
  data-testid="hasanat-token-layer"
>
  <div
    aria-live="polite"
    class="sr-only"
    role="status"
  >
    Hasanāt +3. Total: 128.
  </div>
  <div
    class="pointer-events-none absolute left-0 top-0 will-change-transform"
    style="transform: translate3d(220px, 560px, 0);"
  >
    <div
      aria-hidden="true"
      class="pointer-events-none select-none rounded-full bg-emerald-500/90 px-3 py-1.5 text-xs font-semibold text-white shadow-[0_16px_40px_rgba(16,185,129,0.35)] ring-1 ring-emerald-300/35 backdrop-blur-sm dark:bg-emerald-400/90 dark:text-emerald-950 animate-hasanat-token-float"
      style="opacity: 0; transform: translate(-50%, -50%) scale(0.85); contain: paint; --hasanat-float-distance: 58px; --hasanat-float-duration: 900ms;"
    >
      <div
        class="flex flex-col items-center gap-0.5 leading-none"
      >
        <span
          class="text-base font-semibold tracking-tight"
        >
          +
          1
        </span>
        <span
          class="text-[0.6rem] font-medium uppercase tracking-[0.35em] text-emerald-50/80 dark:text-emerald-900/70"
        >
          Hasanāt
        </span>
      </div>
    </div>
  </div>
  <div
    class="pointer-events-none absolute left-0 top-0 will-change-transform"
    style="transform: translate3d(240px, 580px, 0);"
  >
    <div
      aria-hidden="true"
      class="pointer-events-none select-none rounded-full bg-emerald-500/90 px-3 py-1.5 text-xs font-semibold text-white shadow-[0_16px_40px_rgba(16,185,129,0.35)] ring-1 ring-emerald-300/35 backdrop-blur-sm dark:bg-emerald-400/90 dark:text-emerald-950"
      style="opacity: 0; transform: translate(-50%, -50%) scale(0.85); contain: paint;"
    >
      <div
        class="flex flex-col items-center gap-0.5 leading-none"
      >
        <span
          class="text-base font-semibold tracking-tight"
        >
          +
          3
        </span>
        <span
          class="text-[0.6rem] font-medium uppercase tracking-[0.35em] text-emerald-50/80 dark:text-emerald-900/70"
        >
          Hasanāt
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`FloatingTokensLayer > shows Arabic accent labels when locale is Arabic 1`] = `
<div
  class="pointer-events-none fixed inset-0 z-[70] overflow-visible"
  data-testid="hasanat-token-layer"
>
  <div
    aria-live="polite"
    class="sr-only"
    role="status"
  >
    Hasanāt +2. Total: 128.
  </div>
  <div
    class="pointer-events-none absolute left-0 top-0 will-change-transform"
    style="transform: translate3d(240px, 580px, 0);"
  >
    <div
      aria-hidden="true"
      class="pointer-events-none select-none rounded-full bg-emerald-500/90 px-3 py-1.5 text-xs font-semibold text-white shadow-[0_16px_40px_rgba(16,185,129,0.35)] ring-1 ring-emerald-300/35 backdrop-blur-sm dark:bg-emerald-400/90 dark:text-emerald-950"
      style="opacity: 0; transform: translate(-50%, -50%) scale(0.85); contain: paint; transition: opacity 260ms ease-out;"
    >
      <div
        class="flex flex-col items-center gap-0.5 leading-none"
      >
        <span
          class="text-base font-semibold tracking-tight"
        >
          +
          2
        </span>
        <span
          class="text-[0.6rem] font-medium uppercase tracking-[0.35em] text-emerald-50/80 dark:text-emerald-900/70"
        >
          Hasanāt
        </span>
        <span
          class="text-[0.65rem] font-medium text-emerald-100/80 dark:text-emerald-900/70"
        >
          حسنات
        </span>
      </div>
    </div>
  </div>
</div>
`;
